---
title: "Transcript, Gene and Gene-set network integration"
author:
- name: Izabela Mamede Costa Andrade da Conceição
  affiliation: Departament of Genetics, Ecology and Evolution, Institute of Biological Sciences, Federal University of Minas Gerais (UFMG) - Brazil
  email: iza.mamede@gmail.com
- name: Lucio Rezende Queiroz
  affiliation: Departament of Biochemistry and Immunology, Institute of Biological Sciences, Federal University of Minas Gerais (UFMG) - Brazil
  email: luciorqueiroz@ufmg.br
package: txGeneNetwork
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
vignette: >
  %\VignetteIndexEntry{Transcript, Gene and Gene-set network integration}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
abstract: |
  This workflow is a comprehensive guideline of how to construct Pathway-Gene-Transcript network 
  from transcript expression data using the amazing packs 
  [tidygraph](https://github.com/thomasp85/tidygraph) and [ggraph](https://github.com/thomasp85/ggraph).
keywords: RNA-seq, workflow, network, functional analysis
output:
  BiocStyle::html_document:
    fig_caption: yes
    toc_float: yes
bibliography: references.bib
---

<!--
****
add to description
URL: https://usethis.r-lib.org, https://github.com/r-lib/usethis
BugReports: https://github.com/r-lib/usethis/issues
-->

```{r style, echo=FALSE, results='hide', message=FALSE}
library(BiocStyle)
library(knitr)
knitr::opts_chunk$set(error = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
# knitr::opts_chunk$set(fig.asp = 1)
```

# Workflow version information

**R version**: `r R.version.string`

**Bioconductor version**: `r BiocManager::version()`

**Package**: `r packageVersion("txGeneNetwork")`

# Introduction

<!-- TODO Fazer uma introduçãozinha para a versão da revista -->

<!--

This workflow was motivated by ...

Functional analysis is widely used to ...

-->

# Getting started

Install and run all needed packages

```{r load_deps, echo=FALSE, message=FALSE, warning=FALSE}
# txGeneNetwork:::load_dependencies()
suppressPackageStartupMessages({
  library(ggraph)
  library(tidygraph)
  library(ggforce)
  library(readr)
  library(tidyr)
  library(ggnewscale)
  library(concaveman)
})
```

```{r load_dependencies2}
library(ggforce)
library(readr)
library(tidyr)
library(ggnewscale)
library(tidygraph)
library(ggraph)
library(concaveman)

library(txGeneNetwork)
```

# Input data

<!--
Other Bioconductor packages that can be used to ...
-->

## Quick Start

Your input data must be organized as a .csv with **From** and **To** Columns, for this specific graph your from-to order would go Process -\> Gene and in another line Gene -\> Transcript

| From      | To           |
| --------- | ------------ |
| Process_1 | Gene_1       |
| Gene_1    | Transcript_1 |
| Gene_1    | Transcript_2 |

Other metadata columns which will correspond to the **edges** info can be added and will also be imported, for this type of network we add a Process column, or to which metabolic process that edge belongs to and a Direction column, to show if that transcript is up or down regulated in our example data.

| From      | To           | Process   | Direction |
| --------- | ------------ | --------- | --------: |
| Process_1 | Gene_1       | Process_1 | NA        |
| Gene_1    | Transcript_1 | Process_1 | Up        |
| Gene_1    | Transcript_2 | Process_1 | Down      |

# Plotting

`tidygraph` uses a two tibble format, one for nodes and one for edges and displays it as a **tbl_graph** object, using a tidy manner to display both tibbles together

```{r message=FALSE, warning=FALSE}
example_dataset_path <- system.file("extdata", "example_dataset.csv", package = "txGeneNetwork")
example_dataset <- read_csv(example_dataset_path)
```

The `tbl_graph()` command allows you to directly create a `tbl_graph` object using our .csv table

```{r}
example_tbl_graph <- as_tbl_graph(example_dataset)
example_tbl_graph
```

## Modified ggplot2 syntax

Now that we have the `tbl_graph` object we can start plotting the data. GGraph uses a syntax very similar to ggplot2 and most of the addons used in `ggplot2` can also be used in `ggraph`, like `theme_*()` from `ggthemes` and `geom_*_repel()` from `ggrepel`.
To start we will construct the basic network using the example data and after we add extra information for nodes and for edges.
We will use `geom_node_point()` and `geom_edge_link()` for the basic network.

```{r fig1, fig.height=3, fig.width=5}
example_tbl_graph %>%
  ggraph() +
  geom_node_point() +
  geom_edge_link()
```

We got a message, saying that `ggraph` used *sugiyama* as default layout, that can be changed passing an argument to the `ggraph()` function call.

```{r fig2, fig.height=3, fig.width=3}
example_tbl_graph %>%
  ggraph(layout = "kk") +
  geom_node_point() +
  geom_edge_link()
```

Now we have a network more similar to the final product. 
To modify our `tbl_graph` object and add other variables you can use usual `dplyr` syntax together with the `activate()` function.
The `activate()` will select which of the tibbles you are modifying the nodes or the edges tibble.
Here we add a centrality measure ^[add a reference about network statistics] to the network and size it accordingly using an `aes()` call inside `geom_node_point()`.

```{r fig3, fig.height=3, fig.width=4}
example_tbl_graph %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_node_point(aes(size = centrality)) +
  geom_edge_link()
```

We can also color de edges according to the process they belong to, or to the direction of the transcript expression, using a similar syntax, but now adding an `aes()` call inside `geom_edge_link()`.

```{r fig4, fig.height=3, fig.width=4}
example_tbl_graph %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_edge_link(aes(col = Direction)) +
  geom_node_point(aes(size = centrality))
```

```{r fig5, fig.height=3, fig.width=6}
example_tbl_graph %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_edge_link(aes(col = Group)) +
  geom_node_point(aes(size = centrality))
```

As there are genes which belong to more than one biological process, this is not an adequate process visualization, the best way would be plotting it as individual hulls, but we will get to that down the workflow.

## Adding information on the nodes

For now in our example table we only added aesthetics to the edges, now we will add the transcript_type and the hull aesthetic. First you can extract the nodes table to then modify it using

```{r message=FALSE}
as_tibble(example_tbl_graph %>% activate(nodes))
```

Saving this on an object allows you to save and modify at will your nodes table. Here we load the modified table version

```{r message=FALSE, warning=FALSE}
modified_nodes_path <- system.file("extdata", "modified_nodes.csv", package = "txGeneNetwork")

modified_nodes <- read_csv(modified_nodes_path)
```

We now add the transcript modified information in our `tbl_graph` object and plot it using the `aes(col)`

```{r fig6, fig.height=4.5, fig.width=6.5}
example_tbl_graph %>%
  activate(nodes) %>%
  mutate(Type = modified_nodes$Type) %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_edge_link(aes(col = Direction)) +
  geom_node_point(aes(size = centrality, color = Type))
```

The geom_mark_hull(), the function to add the hull colors as process, does not work well with the `tbl_graph` object due to not being able to add multiple information for the same node in the same color. So the best way to color hulls is to add extra columns representing these sobrepositions and do one geom_mark_hull() call for each.

<!-- TODO create a for loop or map() function to create columns based on the number of Pathways -->

```{r fig7, fig.height=7, fig.width=10}
example_tbl_graph %>%
  activate(nodes) %>%
  mutate(
    Type = modified_nodes$Type,
    Process = modified_nodes$Process_1,
    Process_2 = modified_nodes$Process_2,
    Process_3 = modified_nodes$Process_3
  ) %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_mark_hull(aes(x = x, y = y, fill = Process, color = Process)) +
  geom_mark_hull(aes(x = x, y = y, fill = Process_2, color = Process_2)) +
  geom_mark_hull(aes(x = x, y = y, fill = Process_3, color = Process_3)) +
  geom_edge_link(aes(col = Direction)) +
  new_scale("color") +
  geom_node_point(aes(size = centrality, color = Type))
```

Unfortunately, there is no way to plot this without the NA values due to the tbl_grpah type and how the geom_mark_hull works, so the NA hulls have to be removed _a posteriori_.

## Final touches

Now some final touches like legend size and title

```{r fig8, fig.height=7, fig.width=10}
example_tbl_graph %>%
  activate(nodes) %>%
  mutate(
    Type = modified_nodes$Type,
    Process = modified_nodes$Process_1,
    Process_2 = modified_nodes$Process_2,
    Process_3 = modified_nodes$Process_3
  ) %>%
  mutate(centrality = centrality_power()) %>%
  ggraph(layout = "kk") +
  geom_mark_hull(aes(x = x, y = y, fill = Process, color = Process)) +
  geom_mark_hull(aes(x = x, y = y, fill = Process_2, color = Process_2)) +
  geom_mark_hull(aes(x = x, y = y, fill = Process_3, color = Process_3)) +
  geom_edge_link(aes(col = Direction)) +
  new_scale("color") +
  geom_node_point(aes(size = centrality, color = Type))
```

There, now you have the final network and you only need to save it as .pdf or .csv and remove the NA layer of the hull.

# Session information

```{r}
sessionInfo()
# sessioninfo::session_info()
# xfun::session_info()
```

# References
